manpages/bcopy.8
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH BCOPY 8 "26 November 2009" "Kern Sibbald" "Network backup, recovery and verification"
.\" Please adjust this date whenever revising the manpage.
.\"
.SH NAME
 bcopy \- Bacula's 'Copy from tape'
.SH SYNOPSIS
.B bcopy 
.RI [ options ]
.I input-archive 
.I output-archive
.br
.SH DESCRIPTION
This manual page documents briefly the
.B bcopy
command.
.PP
.\" TeX users may be more comfortable with the \fB<whatever>\fP and
.\" \fI<whatever>\fP escape sequences to invode bold face and italics, 
.\" respectively.
.SH OPTIONS
A summary of options is included below.
.TP
.B \-?
Show version and usage of program.
.TP
.BI \-b\  bootstrap
Specify a bootstrap file.
.TP
.BI \-c\  config
Specify configuration file.
.TP
.BI \-d\  nn
Set debug level to \fInn\fP.
.TP
.BI \-dt
Print timestamp in debug output.
.TP
.BI \-i\  input
Specify input Volume names (separated by '|')
.TP
.BI \-o\  output
Specify output Volume names (separated by '|')
.TP
.BI \-p
Proceed inspite of I/O errors.
.TP
.BI \-w\  directory
Specify working directory (default \fI/tmp\fP).
.TP
.B \-v
Set verbose mode.
.SH SEE ALSO
.BR bls (1),
.BR bextract (1).
.br
.SH AUTHOR
This manual page was written by Jose Luis Tallon
.nh 
<jltallon@adv\-solutions.net>.


ChangeLog
              Changelog on version 5.0.x

General:

Release Version 5.0.0

21Jan10
- Remove double secs in pruning output

20Jan10
- Fix priority mgnt in lock manager
- Fix FD crash when plugin running and cancel given
- Stop backup sooner after cancel
- Make a missing plugin during backup fatal
- Update sample-query.sql
- Use make_catalog_backup.pl by default

19Jan10
- Update makeall and add README to release directory
- Remove debug code
- Add more authors
- Fix possible memory leak
- Fix bscan bug introduced with serialization change
- Add a few status client commands
- Make Ukraine po to compile
- Update po files
- More spec improvements from bug #1455
- Enhance FD status to have level

18Jan10
- Fix logwatch install in bacula.spec. Patch from bug #1454
- Remove /etc/logwatch directories from spec file. Patch supplied by bug 
  #1456
- Don't remove spec files -- fixes bug #1458
- Remove FI sanity check code that does not apply after a VirtualFull
- Make Attr/record FI non-fatal if different -- possibly temp + add new session 
  record types
- Keep serial from overruning with bad data

17Jan10
- Add info message on how to stop the file relocation test

15Jan10
- Add new Path tables to grant and drop scripts
- Fix #1450 about FirstWritten not updated with bscan
- Fix error message in grant_postgresql_privileges

12Jan10
- Fix SQLite table creation script

11Jan10
- Add Job.Reviewed field
- Enable priority checks in lockmanager
- Change DB version from 11 to 12
- Fix Win64 build
- Fix Win32 build

10Jan10
- Fix the subject to something easy to delete
- Turn off messages test -- it creates tons of emails

09Jan10
- Optimize strtok usage in ua_cmds

08Jan10
- Remove old code requiring different storages for migration
- More checking for OpenSSL in configure

06Jan10
- Add make_catalog_backup.pl script that uses env variables and disk file to 
  pass database password for backup
- Modify the sql query to get alphabetical order of clients when selecting the 
  most recent backup for a client

05Jan10
- Remove warning about make_catalog_backup.pl

04Jan10
- Add Ingres configuration
- Fix Ingres ID definition that caused seg fault
- Add commit from Stefan Reddig

03Jan10
- Simplify initgroups definition for AIX
- initgroups definition is ok on aix >= 5

31Dec09
- Ignore scripts/bacula_config
- Remove QWT from configure -- not used

28Dec09
- Revert to configure with C due to old prototypes on some systems
- Make ioctl_req_t test less g++ dependent
- Make ioctl test more exact
- Try to force configure compiles to be more exact
- Rebuild configure to handle Marco's changes
- Clean old configure

27Dec09
- Let --enable-libtool and --disable-libtool determine if we build plugins and 
  drop the extra configure option

26Dec09
- Lets use a xattr specific constant and not the stdio BUFSIZ which is kind of 
  platform dependend.
- Attempt to autoconfigure ioctl_req_t
- Lets not mix initialized stack variable with uninitialized ones in the same 
  definition.
- Added some handling of EPERM and EOPNOTSUPP which seem to be returned when 
  you try to retrieve the system extended attributes a non root user.
- Lets not orphan the buffer returned by extattr_namespace_to_string as it 
  seems by looking at the code it strdups the string it returns so no need to 
  bstrdup it ourself.

25Dec09
- Show that plugins are enabled or not in configure output.

24Dec09
- Plugins are dependable on libtool now so cleaned the Makefiles and the fd 
  plugins are only build when --enable-plugins is set and libtool is used.
- Clarify catalog grant permissions message

23Dec09
- Fix depend for changes to SRC variables.
- Drop the usage of macros within macros as it seems some platforms native 
  makes don't support that. While converting the Makefiles also made things 
  somewhat more uniform in naming the lib objects.

22Dec09
- Add scripts/bacula_config script that displays configure options and results
- Continue to support Batch Mode with PostgreSQL < 8.2

21Dec09
- M_ABORT Bacula If batch insert is turned on when we try to open a connection 
  and thread safe is not enabled
- Try to fix problem on FreeBSD with unsupported extended attributes which 
  gives lots of orphaned buffers as seen in regression testing.

20Dec09
- Fix thread safe detection in ./configure

19Dec09
- Update bweb deb
- Fix breload make_catalog_backup.pl rights

17Dec09
- LSBize Debian init scripts

15Dec09
- Fix #undef of lock manager

13Dec09
- Remove checkout master
- Add more debug for Solaris regression timeouts
- Make SUN C++ compiler stop complaining about enum definition.
- Use %c and %n in bootstrap in default bacula-dir.conf

11Dec09
- remove a ref to sqlite
- Remove ua_prune.c debug code
- Print any traceback or bactrace
- Fix old stuff in Makefile
- Allow deadlock manager on FreeBSD, MacOSX and SunOs
- Add allowcompress-test
- Add test for AllowCompress option

10Dec09
- Eliminate compiler warning
- Rework Mac OS 10.6 mtio.h fix
- Attempt to fix mtio.h problem on Mac 10.6
- Allow mount to wake job waiting for op
- Make wait longer for better recycling
- Fix another volume_lock dlock order problem
- Fix and add ASSERT debug
- Fix compilation problem with lockmgr

09Dec09
- Fix misnamed variable
- Reorder lock_volumes and dlock in SD to avoid race conditions
- - five more updates to manpages options I had forgotten in a corner, - a 
  reformated and edited page for bsmtp - a brand new one for btraceback, based 
  on the Problem Resolution Guide and discussions (to be continued?) on this 
  list.
- Fix compilation pb
- Add mutex some priority info for SD
- Add info in dump about lock priority
- Fix CentOS/RHEL detection
- Fix #defines when lock manager turned off
- Add USE_LOCKMGR_PRIORITY switch to enable new priority management
- Add Macro BTHREAD_MUTEX_PRIORITY(p) and lib/mutex_list.h to manage locks
- Make pthread_mutex_init/destroy compatible with bthread_mutex_t
- Fix SD Vol+dev lock race bug

08Dec09

src/stored/bcopy.c
/*
   Bacula® - The Network Backup Solution

   Copyright (C) 2002-2008 Free Software Foundation Europe e.V.

   The main author of Bacula is Kern Sibbald, with contributions from
   many others, a complete list can be found in the file AUTHORS.
   This program is Free Software; you can redistribute it and/or
   modify it under the terms of version two of the GNU General Public
   License as published by the Free Software Foundation and included
   in the file LICENSE.

   This program is distributed in the hope that it will be useful, but
   WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
   General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
   02110-1301, USA.

   Bacula® is a registered trademark of Kern Sibbald.
   The licensor of Bacula is the Free Software Foundation Europe
   (FSFE), Fiduciary Program, Sumatrastrasse 25, 8006 Zürich,
   Switzerland, email:ftf@fsfeurope.org.
*/
/*
 *
 *  Program to copy a Bacula from one volume to another.
 *
 *   Kern E. Sibbald, October 2002
 *
 *
 *   Version $Id$
 */

#include "bacula.h"
#include "stored.h"

/* Dummy functions */
int generate_daemon_event(JCR *jcr, const char *event) { return 1; }
extern bool parse_sd_config(CONFIG *config, const char *configfile, int exit_code);

/* Forward referenced functions */
static void get_session_record(DEVICE *dev, DEV_RECORD *rec, SESSION_LABEL *sessrec);
static bool record_cb(DCR *dcr, DEV_RECORD *rec);


/* Global variables */
static DEVICE *in_dev = NULL;
static DEVICE *out_dev = NULL;
static JCR *in_jcr;                    /* input jcr */
static JCR *out_jcr;                   /* output jcr */
static BSR *bsr = NULL;
static const char *wd = "/tmp";
static bool list_records = false;
static uint32_t records = 0;
static uint32_t jobs = 0;
static DEV_BLOCK *out_block;
static SESSION_LABEL sessrec;

static CONFIG *config;
#define CONFIG_FILE "bacula-sd.conf"
char *configfile = NULL;
STORES *me = NULL;                    /* our Global resource */
bool forge_on = false;                /* proceed inspite of I/O errors */
pthread_mutex_t device_release_mutex = PTHREAD_MUTEX_INITIALIZER;
pthread_cond_t wait_device_release = PTHREAD_COND_INITIALIZER;


static void usage()
{
   fprintf(stderr, _(
PROG_COPYRIGHT
"\nVersion: %s (%s)\n\n"
"Usage: bcopy [-d debug_level] <input-archive> <output-archive>\n"
"       -b bootstrap      specify a bootstrap file\n"
"       -c <file>         specify a Storage configuration file\n"
"       -d <nn>           set debug level to <nn>\n"
"       -dt               print timestamp in debug output\n"
"       -i                specify input Volume names (separated by |)\n"
"       -o                specify output Volume names (separated by |)\n"
"       -p                proceed inspite of errors\n"
"       -v                verbose\n"
"       -w <dir>          specify working directory (default /tmp)\n"
"       -?                print this message\n\n"), 2002, VERSION, BDATE);
   exit(1);
}

int main (int argc, char *argv[])
{
   int ch;
   char *iVolumeName = NULL;
   char *oVolumeName = NULL;
   bool ignore_label_errors = false;
   bool ok;

   setlocale(LC_ALL, "");
   bindtextdomain("bacula", LOCALEDIR);
   textdomain("bacula");
   init_stack_dump();

   my_name_is(argc, argv, "bcopy");
   lmgr_init_thread();
   init_msg(NULL, NULL);

   while ((ch = getopt(argc, argv, "b:c:d:i:o:pvw:?")) != -1) {
      switch (ch) {
      case 'b':
         bsr = parse_bsr(NULL, optarg);
         break;

      case 'c':                    /* specify config file */
         if (configfile != NULL) {
            free(configfile);
         }
         configfile = bstrdup(optarg);
         break;

      case 'd':                    /* debug level */
         if (*optarg == 't') {
            dbg_timestamp = true;
         } else {
            debug_level = atoi(optarg);
            if (debug_level <= 0) {
               debug_level = 1;
            }
         }
         break;

      case 'i':                    /* input Volume name */
         iVolumeName = optarg;
         break;

      case 'o':                    /* output Volume name */
         oVolumeName = optarg;
         break;

      case 'p':
         ignore_label_errors = true;
         forge_on = true;
         break;

      case 'v':
         verbose++;
         break;

      case 'w':
         wd = optarg;
         break;

      case '?':
      default:
         usage();

      }
   }
   argc -= optind;
   argv += optind;

   if (argc != 2) {
      Pmsg0(0, _("Wrong number of arguments: \n"));
      usage();
   }

   OSDependentInit();

   working_directory = wd;

   if (configfile == NULL) {
      configfile = bstrdup(CONFIG_FILE);
   }

   config = new_config_parser();
   parse_sd_config(config, configfile, M_ERROR_TERM);

   /* Setup and acquire input device for reading */
   Dmsg0(100, "About to setup input jcr\n");
   in_jcr = setup_jcr("bcopy", argv[0], bsr, iVolumeName, 1); /* read device */
   if (!in_jcr) {
      exit(1);
   }
   in_jcr->ignore_label_errors = ignore_label_errors;
   in_dev = in_jcr->dcr->dev;
   if (!in_dev) {
      exit(1);
   }

   /* Setup output device for writing */
   Dmsg0(100, "About to setup output jcr\n");
   out_jcr = setup_jcr("bcopy", argv[1], bsr, oVolumeName, 0); /* no acquire */
   if (!out_jcr) {
      exit(1);
   }
   out_dev = out_jcr->dcr->dev;
   if (!out_dev) {
      exit(1);
   }
   Dmsg0(100, "About to acquire device for writing\n");
   /* For we must now acquire the device for writing */
   out_dev->r_dlock();
   if (out_dev->open(out_jcr->dcr, OPEN_READ_WRITE) < 0) {
      Emsg1(M_FATAL, 0, _("dev open failed: %s\n"), out_dev->errmsg);
      out_dev->dunlock();
      exit(1);
   }
   out_dev->dunlock();
   if (!acquire_device_for_append(out_jcr->dcr)) {
      free_jcr(in_jcr);
      exit(1);
   }
   out_block = out_jcr->dcr->block;

   ok = read_records(in_jcr->dcr, record_cb, mount_next_read_volume);

   if (ok || out_dev->can_write()) {
      if (!write_block_to_device(out_jcr->dcr)) {
         Pmsg0(000, _("Write of last block failed.\n"));
      }
   }

   Pmsg2(000, _("%u Jobs copied. %u records copied.\n"), jobs, records);

   free_jcr(in_jcr);
   free_jcr(out_jcr);

   in_dev->term();
   out_dev->term();
   return 0;
}


/*
 * read_records() calls back here for each record it gets
 */
static bool record_cb(DCR *in_dcr, DEV_RECORD *rec)
{
   if (list_records) {
      Pmsg5(000, _("Record: SessId=%u SessTim=%u FileIndex=%d Stream=%d len=%u\n"),
            rec->VolSessionId, rec->VolSessionTime, rec->FileIndex,
            rec->Stream, rec->data_len);
   }
   /*
    * Check for Start or End of Session Record
    *
    */
   if (rec->FileIndex < 0) {
      get_session_record(in_dcr->dev, rec, &sessrec);

      if (verbose > 1) {
         dump_label_record(in_dcr->dev, rec, 1);
      }
      switch (rec->FileIndex) {
      case PRE_LABEL:
         Pmsg0(000, _("Volume is prelabeled. This volume cannot be copied.\n"));
         return false;
      case VOL_LABEL:
         Pmsg0(000, _("Volume label not copied.\n"));
         return true;
      case SOS_LABEL:
         if (bsr && rec->match_stat < 1) {
            /* Skipping record, because does not match BSR filter */
            if (verbose) {
             Pmsg0(-1, _("Copy skipped. Record does not match BSR filter.\n"));
            }
         } else {
            jobs++;
         }
         break;
      case EOS_LABEL:
         if (bsr && rec->match_stat < 1) {
            /* Skipping record, because does not match BSR filter */
           return true;
        }
         while (!write_record_to_block(out_block, rec)) {
            Dmsg2(150, "!write_record_to_block data_len=%d rem=%d\n", rec->data_len,
                       rec->remainder);
            if (!write_block_to_device(out_jcr->dcr)) {
               Dmsg2(90, "Got write_block_to_dev error on device %s: ERR=%s\n",
                  out_dev->print_name(), out_dev->bstrerror());
               Jmsg(out_jcr, M_FATAL, 0, _("Cannot fixup device error. %s\n"),
                     out_dev->bstrerror());
               return false;
            }
         }
         if (!write_block_to_device(out_jcr->dcr)) {
            Dmsg2(90, "Got write_block_to_dev error on device %s: ERR=%s\n",
               out_dev->print_name(), out_dev->bstrerror());
            Jmsg(out_jcr, M_FATAL, 0, _("Cannot fixup device error. %s\n"),
                  out_dev->bstrerror());
            return false;
         }
         return true;
      case EOM_LABEL:
         Pmsg0(000, _("EOM label not copied.\n"));
         return true;
      case EOT_LABEL:              /* end of all tapes */
         Pmsg0(000, _("EOT label not copied.\n"));
         return true;
      default:
         return true;
      }
   }

   /*  Write record */
   if (bsr && rec->match_stat < 1) {
      /* Skipping record, because does not match BSR filter */
      return true;
   }
   records++;
   while (!write_record_to_block(out_block, rec)) {
      Dmsg2(150, "!write_record_to_block data_len=%d rem=%d\n", rec->data_len,
                 rec->remainder);
      if (!write_block_to_device(out_jcr->dcr)) {
         Dmsg2(90, "Got write_block_to_dev error on device %s: ERR=%s\n",
            out_dev->print_name(), out_dev->bstrerror());
         Jmsg(out_jcr, M_FATAL, 0, _("Cannot fixup device error. %s\n"),
               out_dev->bstrerror());
         return false;
      }
   }
   return true;
}

static void get_session_record(DEVICE *dev, DEV_RECORD *rec, SESSION_LABEL *sessrec)
{
   const char *rtype;
   memset(sessrec, 0, sizeof(sessrec));
   switch (rec->FileIndex) {
   case PRE_LABEL:
      rtype = _("Fresh Volume Label");
      break;
   case VOL_LABEL:
      rtype = _("Volume Label");
      unser_volume_label(dev, rec);
      break;
   case SOS_LABEL:
      rtype = _("Begin Job Session");
      unser_session_label(sessrec, rec);
      break;
   case EOS_LABEL:
      rtype = _("End Job Session");
      unser_session_label(sessrec, rec);
      break;
   case 0:
   case EOM_LABEL:
      rtype = _("End of Medium");
      break;
   default:
      rtype = _("Unknown");
      break;
   }
   Dmsg5(10, "%s Record: VolSessionId=%d VolSessionTime=%d JobId=%d DataLen=%d\n",
         rtype, rec->VolSessionId, rec->VolSessionTime, rec->Stream, rec->data_len);
   if (verbose) {
      Pmsg5(-1, _("%s Record: VolSessionId=%d VolSessionTime=%d JobId=%d DataLen=%d\n"),
            rtype, rec->VolSessionId, rec->VolSessionTime, rec->Stream, rec->data_len);
   }
}


/* Dummies to replace askdir.c */
bool    dir_find_next_appendable_volume(DCR *dcr) { return 1;}
bool    dir_update_volume_info(DCR *dcr, bool relabel, bool update_LastWritten) { return 1; }
bool    dir_create_jobmedia_record(DCR *dcr, bool zero) { return 1; }
bool    dir_ask_sysop_to_create_appendable_volume(DCR *dcr) { return 1; }
bool    dir_update_file_attributes(DCR *dcr, DEV_RECORD *rec) { return 1;}
bool    dir_send_job_status(JCR *jcr) {return 1;}


bool dir_ask_sysop_to_mount_volume(DCR *dcr, int /*mode*/)
{
   DEVICE *dev = dcr->dev;
   fprintf(stderr, _("Mount Volume \"%s\" on device %s and press return when ready: "),
      dcr->VolumeName, dev->print_name());
   dev->close();
   getchar();
   return true;
}

bool dir_get_volume_info(DCR *dcr, enum get_vol_info_rw  writing)
{
   Dmsg0(100, "Fake dir_get_volume_info\n");
   bstrncpy(dcr->VolCatInfo.VolCatName, dcr->VolumeName, sizeof(dcr->VolCatInfo.VolCatName));
   dcr->VolCatInfo.VolCatParts = find_num_dvd_parts(dcr);
   Dmsg2(500, "Vol=%s num_parts=%d\n", dcr->VolCatInfo.VolCatName, dcr->VolCatInfo.VolCatParts);
   return 1;
}
